### PR TITLE
Refactor dimension-related API functions

### PR DESCRIPTION
**Pull request by erimatnor**
The functions for adding and updating dimensions have been refactored
in C to:

- improve usage of proper error codes
- make  messages that better conform with the PostgreSQL standard.
- improve security by avoiding that lots of code run under SECURITY DEFINER

A new if_not_exists option has also been added to add_dimension() and
a the number of partitions can now be set using the new
set_num_partitions() function.

A bug in the validation of smallint time intervals has been fixed. The
previous code didn't check for intervals > 0 and smallint intervals
accepted values up to UINT16_MAX instead of INT16_MAX.

### CODE REVIEW

**Comment by erimatnor** [comment]
@cevian @RobAtticus This PR is also a first step in cleaning up error handling, with more consistent messages (PostgreSQL conforming), proper error codes. Please pay extra attention to error message and return codes when reviewing. I would appreciate feedback on those as they are user-facing.

File: sql/ddl_internal.sql [hunk 1]
```diff
@@ -65,6 +66,22 @@ BEGIN
         USING ERRCODE ='IO101';
     END IF;
 
+    -- Pre-validate dimensions to ensure we do not unnecessarily
+    -- increment the hypertable ID
+    PERFORM _timescaledb_internal.validate_dimension(main_table,
```
Comment (erimatnor) [inline sql/ddl_internal.sql]: FWIW, I have another PR in the pipeline that will clean up `create_hypertable()`, which will make this stuff a lot nicer.

File: src/dimension.h [hunk 2]
```diff
@@ -64,15 +64,17 @@ typedef struct Point
 #define POINT_SIZE(cardinality)							\
 	(sizeof(Point) + (sizeof(int64) * (cardinality)))
 
+#define DEFAULT_CHUNK_TIME_INTERVAL (USECS_PER_DAY * 30)	/* 1 month */
```
Comment (RobAtticus) [inline src/dimension.h]: I wonder if now is a good time to switch from 30 days to 1 day or something smaller?

File: src/dimension.c [hunk 3]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
...
+		.interval_type = PG_ARGISNULL(3) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 3),
+		.partitioning_func = PG_ARGISNULL(4) ? InvalidOid : PG_GETARG_OID(4),
+		.if_not_exists = PG_ARGISNULL(5) ? false : PG_GETARG_BOOL(5),
+		.ht = hypertable_cache_get_entry(hcache, PG_GETARG_OID(0)),
+		.show_warnings = true,
+	};
+
+	hypertable_permissions_check(info.table_relid, GetUserId());
+
+	if (!hypertable_lock_tuple_simple(info.table_relid))
```
Comment (cevian) [inline src/dimension.c]: I am not quite sure why this lock is necessary. Why doesn't this code simply use `hypertable_increment_num_dimensions` instead of `hypertable_set_num_dimensions` and then not have to deal with this locking?

File: test/expected/create_chunks.out [hunk 4]
```diff
@@ -200,14 +205,14 @@ ORDER BY d.id;
 \set ON_ERROR_STOP 0
 -- should fail since time column is an int
 SELECT set_chunk_time_interval('chunk_test', INTERVAL '1 minute');
-ERROR:  chunk_time_interval needs to be an integer type for SMALLINT, INTEGER, and BIGINT time columns
+ERROR:  invalid interval: must be an integer type for integer dimensions
```
Comment (cevian) [inline test/expected/create_chunks.out]: May be good to add dimension column name to all of these.

File: test/expected/create_hypertable.out [hunk 5]
```diff
@@ -162,31 +162,39 @@ NOTICE:  Adding NOT NULL constraint to time column time (NULL time values not al
...
 SELECT add_dimension('test_schema.test_table', 'nope', 2);
 ERROR:  column "nope" does not exist
 --adding the same dimension twice should fail
 select add_dimension('test_schema.test_table', 'location', 2);
-ERROR:  A dimension on column "location" already exists
+ERROR:  column "location" is already a dimension
 --adding dimension with both number_partitions and interval_length should fail
 select add_dimension('test_schema.test_table', 'id2', number_partitions => 2, interval_length => 1000);
-ERROR:  Cannot specify both interval and number of slices/partitions for a single dimension
+ERROR:  a dimension must specify either the number of partitions or an interval
```
Comment (cevian) [inline test/expected/create_hypertable.out]: add `but not both`?

File: test/expected/ddl_errors.out [hunk 6]
```diff
@@ -10,7 +10,7 @@ ERROR:  relation "public.Hypertable_1_mispelled" does not exist at character 33
 SELECT * FROM create_hypertable('"public"."Hypertable_1"', 'time_mispelled', 'Device_id', 2, chunk_time_interval=>_timescaledb_internal.interval_to_usec('1 month'));
 ERROR:  column "time_mispelled" does not exist
 SELECT * FROM create_hypertable('"public"."Hypertable_1"', 'Device_id', 'Device_id', 2, chunk_time_interval=>_timescaledb_internal.interval_to_usec('1 month'));
-ERROR:  illegal type for column "Device_id": text
+ERROR:  invalid dimension type: must be an integer, date or timestamp
```
Comment (cevian) [inline test/expected/ddl_errors.out]: Definitely good to add named of dimension here.

File: sql/ddl_api.sql [hunk 7]
```diff
@@ -194,50 +135,15 @@ $BODY$;
...
-    time_type := _timescaledb_internal.dimension_type(main_table, time_dimension.column_name, true);
-    chunk_time_interval_actual := _timescaledb_internal.time_interval_specification_to_internal_with_default_time(
-        time_type, chunk_time_interval, 'chunk_time_interval');
-
-    UPDATE _timescaledb_catalog.dimension d
-    SET interval_length = chunk_time_interval_actual
-    WHERE time_dimension.id = d.id;
-END
-$BODY$;
+CREATE OR REPLACE FUNCTION  set_num_partitions(
```
Comment (cevian) [inline sql/ddl_api.sql]: Should this be `set_number_partitions` to be the same as the parameter name in create_hypertable?

File: test/expected/timestamp.out [hunk 8]
```diff
@@ -595,52 +595,3 @@ FROM unnest(ARRAY[
  11-09-2017 | 11-09-2017
 (4 rows)
 
-------------------------------------
```
Comment (cevian) [inline test/expected/timestamp.out]: We should still be testing these conversions somewhere

File: src/dimension.c [hunk 9]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
...
+		{
+			if (!info->if_not_exists)
+				ereport(ERROR,
+						(errcode(ERRCODE_IO_DUPLICATE_DIMENSION),
+						 errmsg("column \"%s\" is already a dimension",
+								NameStr(*info->colname))));
+
+			info->skip = true;
+
+			ereport(NOTICE,
```
Comment (cevian) [inline src/dimension.c]: do NOTICES come with error codes?

File: src/dimension.h (hunk 2 above)
Comment (erimatnor) [reply src/dimension.h]: Maybe, but perhaps we should do this in a separate PR to make the switch more explicit?

File: src/dimension.c [hunk 10]
```diff
@@ -379,16 +401,72 @@ dimension_tuple_update(TupleInfo *ti, void *data)
...
+	values[Anum_dimension_column_name - 1] = NameGetDatum(colname);
+	values[Anum_dimension_column_type - 1] = ObjectIdGetDatum(coltype);
+
+	if (OidIsValid(partitioning_func))
+	{
+		Form_pg_proc procform = get_procform(partitioning_func);
+		const char *nspname = get_namespace_name(procform->pronamespace);
+
+		values[Anum_dimension_partitioning_func - 1] = NameGetDatum(&procform->proname);
+		values[Anum_dimension_partitioning_func_schema - 1] = CStringGetDatum(nspname);
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: So, yes, I can add the conversion. However, I actually intentionally did it this way since `pronamespace` is read from `pg_catalog` and is thus guaranteed to be within `NAMEDATALEN`. Conversion is thus unnecessary. Besides, if it wouldn't be, the scan would just fail. How do we want to deal with these cases?

File: src/dimension.c [hunk 11]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
 
 	return p;
 }
+
+static inline int64
+interval_to_usec(Interval *interval)
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: So, my reasoning is that since this function is only used here, it makes sense to have it here where it has the proper context. If you feel strongly, I can move it to `utils.c`. That said, I think our time-related functions in `utils.c` is a bit of a mess right now. I'd like to refactor that into its own file at some point.

File: src/dimension.c [hunk 12]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
...
+#define IS_VALID_OPEN_DIM_TYPE(type)					\
+	(IS_INTEGER_TYPE(type) || IS_TIMESTAMP_TYPE(type))
+
+#define INT_TYPE_MAX(type)												\
+	(int64)((type == INT2OID) ? INT16_MAX : ((type == INT4OID) ? INT32_MAX : INT64_MAX))
+
+static int64
+get_validated_integer_interval(Oid coltype, int64 value, bool show_warnings)
+{
+	if ((coltype == INT2OID && value > INT16_MAX) ||
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: Yeah, that makes sense. Added `INT_TYPE_MAX` after having written this part of the code.

File: src/dimension.c [hunk 13]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
...
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("hypertable \"%s\" has multiple %s dimensions",
+							get_rel_name(table_relid),
+							dimtype == DIMENSION_TYPE_OPEN ? "time" : "space"),
+					 errhint("An explicit dimension name needs to be specified")));
+
+		dim = hyperspace_get_dimension(ht->space, dimtype, 0);
+	}
+	else
+		dim = hyperspace_get_dimension_by_name(ht->space, dimtype, NameStr(*dimname));
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: Not sure what you mean. Do you want an `Assert()`? The type is given as argument when retrieving the dimension. Do you want to test that `hyperspace_get_dimension()` doesn't have a bug? In that case I think we should unit test that. But I am not sure why this function should be singled out since it is quite simple?

File: src/dimension.c [hunk 14]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
...
+	int16		num_slices;
+
+	hypertable_permissions_check(table_relid, GetUserId());
+
+	if (PG_ARGISNULL(1) || num_slices_arg < 1 || num_slices_arg > INT16_MAX)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid number of partitions")));
+
+	num_slices = num_slices_arg & 0xffff;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: So, the in-argument for the function is an `INTEGER` (32 bits) although num_slices is an `SMALLINT` in our catalog. We use an `INTEGER` argument avoid a cast when calling the API function. The value of the `num_slices_arg` `int32` is verified to be within the bounds of an `int16`, then it can be safely converted to an `int16`. Without this stuff, the compiler would complain as `dimension_update()` and the catalog table takes an `int16`. I could do a C-style cast if you prefer that.

File: src/dimension.c (hunk 3 above)
Comment (erimatnor) [reply src/dimension.c]: Yeah so that is one option I certainly considered. A couple of reasons:

1. I tried to follow the old `plpgsql` code that took a row-level lock on the hypertable tuple.
2. We have a `CHECK` constraint on the `hypertable` table that says `num_dimensions > 0`, so `create_hypertable` must set it to at least 1. Therefore, it is not possible to use `hypertable_increment_num_dimensions` since it is not possible to know when `num_dimensions` has already been incremented (e.g., when called from `create_hypertable`) and when not (e.g., when this function is called explicitly). (Still, not sure that the catalog update API enforces check constraints, but makes sense to follow). The old code used a boolean to distinguish between these two cases, which seemed a bit ugly and unnecessary (especially if exposed in a public user-facing API).
3. An explicit row-level lock here makes it easier to bail out with a proper error message. (Although I guess we can take a table-level `RowExclusiveLock` already here)
4. `hypertable_increment_num_dimensions` and `hypertable_set_num_dimensions` take a table-level `RowExclusiveLock` and if the update fails, a general locking error ("tuple already updated") will be thrown that is more obscure. We could capture this error in a `TRY()` block an display a more appropriate error message though.

Open for suggestions though...

File: sql/ddl_api.sql (hunk 7 above)
Comment (erimatnor) [reply sql/ddl_api.sql]: I think that name sounds somewhat awkward, but can change.

File: src/dimension.c [hunk 15]
```diff
@@ -379,16 +401,72 @@ dimension_tuple_update(TupleInfo *ti, void *data)
...
+
+	if (OidIsValid(partitioning_func))
+	{
+		Form_pg_proc procform = get_procform(partitioning_func);
+		const char *nspname = get_namespace_name(procform->pronamespace);
+
+		values[Anum_dimension_partitioning_func - 1] = NameGetDatum(&procform->proname);
+		values[Anum_dimension_partitioning_func_schema - 1] = CStringGetDatum(nspname);
+		values[Anum_dimension_num_slices - 1] = Int16GetDatum(num_slices);
+		values[Anum_dimension_aligned - 1] = false;
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: Only found one place below.

File: sql/ddl_internal.sql [hunk 16]
```diff
@@ -571,6 +448,25 @@ BEGIN
 END
 $BODY$;
 
+CREATE OR REPLACE FUNCTION _timescaledb_internal.add_dimension(
+    main_table               REGCLASS,
+    column_name              NAME,
+    num_slices               INTEGER = NULL,
+    interval_length          anyelement = NULL::BIGINT,
```
Comment (erimatnor) [inline sql/ddl_internal.sql]: Should we rename this parameter to `chunk_time_interval` to be consistent with `create_hypertable()` and `set_chunk_time_interval()`?

File: src/dimension.c [hunk 17]
```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
...
+		ereport(WARNING,
+				(errcode(ERRCODE_AMBIGUOUS_PARAMETER),
+				 errmsg("unexpected interval: smaller than one second"),
+				 errhint("The interval is specified in microseconds")));
+
+	return value;
+}
+
+static int64
+dimension_interval_to_internal(Oid coltype, Oid valuetype, Datum value, bool show_warnings)
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: So, there's only one "interval" that this may concern, e.g.:

```
add_dimension('foo', 'time', interval_length => interval '1 day');
```
 Figured it would be clear that it would concern `time` in this case. Space dimensions don't have "intervals". But I can add if you think it makes a big difference.

File: src/dimension.c (hunk 9 above)
Comment (erimatnor) [reply src/dimension.c]: Good question... Probably not. How would you check the error code if we do not return here and/or have multiple notices?

File: src/dimension.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/dimension.c]: Yes I think a c-style cast is more clear.

File: src/hypertable.c [hunk 18]
```diff
@@ -196,7 +213,120 @@ hypertable_update_form(FormData_hypertable *update)
...
+		case HeapTupleMayBeUpdated:
+			/* successfully locked */
+			return true;
+		case HeapTupleUpdated:
+			ereport(ERROR,
+					(errcode(ERRCODE_IO_HYPERTABLE_NOT_EXIST),
+					 errmsg("hypertable \"%s\" has already been updated by another transaction",
+							get_rel_name(table_relid)),
+					 errhint("Retry the operation again")));
+		case HeapTupleBeingUpdated:
```
[the remark this answers is not in the document]
Comment (cevian) [reply src/hypertable.c]: I am just wondering if these two cases should throw the same error as HeapTupleUpdated. Not sure why some cases throw an error and some don't.

File: test/expected/timestamp.out (hunk 8 above)
Comment (cevian) [reply test/expected/timestamp.out]: Yes but the associated conversion still does and should be tested.

File: sql/ddl_internal.sql (hunk 16 above)
Comment (cevian) [reply sql/ddl_internal.sql]: yes

File: src/dimension.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/dimension.c]: Or was this about muting Valgrind warnings? Anyway, I'll fix.

File: src/hypertable.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/hypertable.c]: I throws an error were I call this function. I wanted the ability for someone using this function to be able to customize the error messages for these failure case to account for context.

File: test/expected/timestamp.out (hunk 8 above)
Comment (erimatnor) [reply test/expected/timestamp.out]: Are you suggesting we expose these functions publicly for the sake of testing? I don't mind testing but why single these functions out compared to a lot of other functions that we don't expose?

Comment (cevian) [reply test/expected/timestamp.out]: We don't necessarily need to expose it but at least test all the edge cases by using set_interval or whatnot. These time conversion functions have had issues multiple times before so I think its a good idea to be preemptive and test these code paths.

Comment (erimatnor) [reply test/expected/timestamp.out]: Ok, managed to add these back by exposing the function. Also figured out a way to only add the SQL part in the tests

**Comment by erimatnor** [comment]
@cevian @RobAtticus should be ready for another round of reviewing

File: src/dimension.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/dimension.c]: ahh didn't see the dimtype arg

File: src/dimension.c (hunk 3 above)
Comment (cevian) [reply src/dimension.c]: Ahh ok. The `num_dimensions > 0` check constraint makes this more clear and persuasive. Can we add in a comment?

File: src/dimension.h (hunk 2 above)
Comment (cevian) [reply src/dimension.h]: Def a separate PR would be preferred IMHO

File: src/hypertable.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (cevian) [reply src/hypertable.c]: I still don't understand why one case throws an error while others return false. Can you explain?

File: test/expected/timestamp.out (hunk 8 above)
Comment (cevian) [reply test/expected/timestamp.out]: I really like this!

File: test/runner.sh [hunk 19]
```diff
@@ -57,4 +61,5 @@ ${PSQL} -U ${TEST_PGUSER} \
      -v ROLE_SUPERUSER=${TEST_ROLE_SUPERUSER} \
      -v ROLE_DEFAULT_PERM_USER=${TEST_ROLE_DEFAULT_PERM_USER} \
      -v ROLE_DEFAULT_PERM_USER_2=${TEST_ROLE_DEFAULT_PERM_USER_2} \
+     -v EXTENSION_LIB="'timescaledb-${EXT_VERSION}'" \
```
Comment (cevian) [inline test/runner.sh]: Why is this needed? Does this bypass the loader and thus make the loader test useless?

File: src/hypertable.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/hypertable.c]: So, at least `WouldBlock` seems like an option on the callers part? So, you might want to handle that error separately. I can throw errors in all failed cases for now if that makes things clearer.

[the remark this answers is not in the document]
Comment (cevian) [reply src/hypertable.c]: I would prefer that but will leave up to you
